### DISCUSSION TITLE
Type inference for Span<T> and co.

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#5842] [Language Ideas] [answered]
For sure this has been asked before but I couldn't find some reference.
While refactoring my methods to `Span<T>` I originally has following method signature

`void GetData<T>(T[] target);`

which copied the data from some internal data into the `target` data. The type could be inferred so there was no need to write

```cs
float[] result = new float[100];
GetData<float>(result);

```

When switching the method signature to `Span<T>`

`void GetData<T>(Span<T> target);`

type inference does not work anymore because T is an interior type now and I had to reworks hundreds of usages, I can understand agree that it cannot be inferred at the moment. Since I have reworked it now it is not much of importance anymore, but for others might have the same issues . Nevertheless I would be nice if this could be made possible. At least for `Span<T>` and `Memory<T>` and of course its read-only versions.

**BUT**: I personally dislike special solutions for special types. So I would be very much against having it on ´Span<T>` and `Memory<T>` only ;-) 

I would appreciate a more generic approach asnd would be curious if there have been ideas on this topic?

**Comment by pinkfloydx33** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/3916
https://github.com/dotnet/csharplang/discussions/2067

**Reply by msedi** [reply]
@pinkfloydx33: Thanks. But do both #3916 and #2067 really describe the same problem? The examples show the type directly in the argument list, but the C# resolution "does not work". In my scenario the problem is that the `type T` is not directly in the argument list, but one level deeper in `Span<T>`.

**Reply by 333fred** [reply]
@msedi absent more details, I believe they are the same issue. They're about how a conversion (from array to Span in this case) doesn't infer types like you want it to.
